kotlinx-coroutines-core/common/src/channels/Channels.common.kt

```diff
@@ -70,23 +104,83 @@ public inline fun <E, R> ReceiveChannel<E>.consume(block: ReceiveChannel<E>.() -
+ * channel's buffer,
+ * or an early return from [action] happens.
+ * If the [action] finishes with an exception, that exception will be used for cancelling the channel and rethrown.
+ * If the channel is [closed][SendChannel.close] with a cause, this cause will be rethrown from [consumeEach].
+ *
+ * When the channel does not need to be closed after iterating over its elements,
+ * a regular `for` loop (`for (element in channel)`) should be used instead.
  *
  * The operation is _terminal_.
  * This function [consumes][ReceiveChannel.consume] all elements of the original [ReceiveChannel].
+ *
+ * This function is useful in cases when this channel is only expected to have a single consumer that decides when
+ * the producer may stop and ensures that the elements that were sent do get processed.
+ * Example:
+ *
+ * ```
+ * fun Int.cleanup() { println("cleaning up $this") }
+ * val channel = Channel<Int>(1, onUndeliveredElement = Int::cleanup)
+ * // Launch several procedures that create values
+ * repeat(5) {
+ *     launch(Dispatchers.Default) {
+ *         while (true) {
+ *             val x = Random.nextInt(40, 50)
+ *             println("Generating $x")
+ *             channel.send(x)
+ *         }
+ *     }
+ * }
+ * // Launch the exclusive consumer
+ * launch(Dispatchers.Default) {
+ *     channel.consumeEach {
+ *         if (it == 42) {
+ *             println("Found the answer")
+ *             return@launch
+ *         } else it.cleanup()
+ *     }
+ * }
+ * ```
+ *
+ * In this example, all ten values created by the producer coroutines will be processed:
+ * while the single consumer is active, it will receive all the elements, but once it exits,
+ * the values that can no longer be delivered will be passed to the `Int.cleanup` handler.
  */
 public suspend inline fun <E> ReceiveChannel<E>.consumeEach(action: (E) -> Unit): Unit =
     consume {
         for (e in this) action(e)
     }
 
 /**
- * Returns a [List] containing all elements.
+ * Returns a [List] containing all the elements sent to this channel, preserving their order.
+ *
+ * This function will attempt to receive elements and put them into the list until the channel is
+ * [closed][SendChannel.close].
+ * Calling [toList] without closing the channel is always incorrect:
+ * - It will suspend indefinitely if the channel is not closed, but no new elements arrive.
+ * - If new elements do arrive and the channel is not eventually closed, [toList] will use more and more memory
+ *   until exhausting it.
+ *
+ * If the channel is [closed][SendChannel.close] with a cause, [toList] will rethrow that cause.
  *
  * The operation is _terminal_.
  * This function [consumes][ReceiveChannel.consume] all elements of the original [ReceiveChannel].
+ *
+ * Example:
+ * ```
+ * val values = listOf(1, 5, 2, 9, 3, 3, 1)
+ * val channel = Channel<Int>()
+ * GlobalScope.launch {
```

Comment:
Let's use `produce` for that? 
Esp. here, where `finally` block is required otherwise

Reply:
I found `produce` too high-level and thought that using it may detract from the point of the sample. Is it wrong? Is `produce` common knowledge and part of any coroutine user's toolkit?

Reply:
`produce` is more high-level, but it's also the idiomatic way to create a producer coroutine, so it's worth mentioning. It's true that it may overwhelm the reader, but we can probably work around this by adding extensive comments.